parser/src/main/scala/org/bykn/bosatsu/parser/Parser.scala

```diff
@@ -0,0 +1,938 @@
+}
+
+object Parser extends ParserInstances {
+  sealed abstract class Expectation {
+    import Expectation._
+
+    def offset: Int
+    def toError: Error = Error.MissedExpectation(this)
+
+    def translateOffset(inc: Int): Expectation = {
+      this match {
+        case Str(o, str) => Str(o + inc, str)
+        case InRange(o, l, c) => InRange(o + inc, l, c)
+        case Failure(o) => Failure(o + inc)
+        case StartOfString(o) => StartOfString(o + inc)
+        case EndOfString(o, length) => EndOfString(o + inc, length)
+        case Length(o, expected, actual) => Length(o + inc, expected, actual)
+      }
+    }
+  }
+
+  object Expectation {
+    case class Str(offset: Int, str: String) extends Expectation
+    // expected a character in a given range
+    case class InRange(offset: Int, lower: Char, upper: Char) extends Expectation
+    case class Failure(offset: Int) extends Expectation
+    case class StartOfString(offset: Int) extends Expectation
+    case class EndOfString(offset: Int, length: Int) extends Expectation
+    case class Length(offset: Int, expected: Int, actual: Int) extends Expectation
+  }
+
+  sealed abstract class Error {
+    def expected: NonEmptyList[Expectation]
+    def offsets: NonEmptyList[Int] =
+      expected.map(_.offset).distinct
+
+    def translateOffset(inc: Int): Error =
+      this match {
+        case Error.MissedExpectation(x) => Error.MissedExpectation(x.translateOffset(inc))
+        case Error.Combined(es) => Error.Combined(es.map(_.translateOffset(inc)))
+      }
+  }
+  object Error {
+    case class MissedExpectation(expectation: Expectation) extends Error {
+      def expected = NonEmptyList(expectation, Nil)
+    }
+
+    case class Combined(errors: NonEmptyList[Error]) extends Error {
+      def expected = errors.flatMap(_.expected)
+    }
+
+    def combined(errors: NonEmptyList[Error]): Error =
+      if (errors.tail.isEmpty) errors.head
+      else Combined(errors)
+  }
+
+  final class With1[+A](val parser: Parser[A]) extends AnyVal {
+    def ~[B](that: Parser1[B]): Parser1[(A, B)] =
+      Parser.product01(parser, that)
+
+    /**
+     * This is the usual monadic composition, but you
+     * should much prefer to use ~ or Apply.product, *>, <*, etc
+     * if you can since it is much more efficient. This
+     * has to call fn on each parse, which could be a lot
+     * of extra work is you already know the result as is
+     * the case for ~
+     */
+     def flatMap[B](fn: A => Parser1[B]): Parser1[B] =
+        Parser.flatMap01(parser)(fn)
+  }
+
+  def pure[A](a: A): Parser[A] =
+    Impl.Pure(a)
+
+  /**
+   * this is an error if string is empty
+   */
+  def expect(str: String): Parser1[Unit] =
+    Impl.Expect(str)
+
+  def oneOf1[A](parsers: List[Parser1[A]]): Parser1[A] = {
+    @annotation.tailrec
+    def flatten(ls: List[Parser1[A]], acc: List[Parser1[A]]): Parser1[A] =
+      ls match {
+        case Nil =>
+          acc match {
```

Comment:
We could call distinct here I think, which would help when composing possibly overlapping unions.